server/server.go

```diff
@@ -11,21 +12,50 @@ import (
 	"github.com/openfluxcd/controller-manager/storage"
 )
 
-type StartServer func(path, address string) error
-
-// InitializeStorage creates a storage and returns the means to launch a file server to serve created Artifacts.
-func InitializeStorage(c client.Client, scheme *runtime.Scheme, path, storageAddress string, artifactRetentionTTL time.Duration, artifactRetentionRecords int) (StartServer, *storage.Storage, error) {
+// NewStorage creates a storage and returns the means to launch a file server to serve created Artifacts.
+func NewStorage(c client.Client, scheme *runtime.Scheme, path, storageAddress string, artifactRetentionTTL time.Duration, artifactRetentionRecords int) (*storage.Storage, error) {
 	stg, err := storage.NewStorage(c, scheme, path, storageAddress, artifactRetentionTTL, artifactRetentionRecords)
 	if err != nil {
-		return nil, nil, fmt.Errorf("error initializing storage: %v", err)
+		return nil, fmt.Errorf("error initializing storage: %v", err)
 	}
 
-	return startFileServer, stg, nil
+	return stg, nil
+}
+
+type ArtifactServer struct {
+	*http.Server
+	timeout time.Duration
 }
 
-func startFileServer(path string, address string) error {
+func NewArtifactServer(path string, address string, timeout time.Duration) (*ArtifactServer, error) {
 	fs := http.FileServer(http.Dir(path))
 	mux := http.NewServeMux()
 	mux.Handle("/", fs)
-	return http.ListenAndServe(address, mux)
+
+	s := &http.Server{
+		Addr:    address,
+		Handler: mux,
+	}
+	as := &ArtifactServer{
+		Server:  s,
+		timeout: timeout,
+	}
+	return as, nil
+}
+
+func (s *ArtifactServer) Start(ctx context.Context) error {
+	serverErr := make(chan error, 1)
+	go func() {
+		serverErr <- s.ListenAndServe()
+	}()
+	var err error
+	var cancel context.CancelFunc
+	select {
+	case <-ctx.Done():
+		ctx, cancel = context.WithTimeout(context.Background(), s.timeout)
+		defer cancel()
+		err = s.Shutdown(ctx)
+	case err = <-serverErr:
+	}
+	return err
```

Comment:
As a side-note, you might want to add a little check here saying:
```go
if errors.Is(err, http.ErrServerClosed) { return nil }
```

Or something like that, because it could throw off the enclosing code. Why? Because there is a chance that s.Shutdown races and tries closing an already closing server. In which case, the "idiomatic" Go way is usually, either just log, or completely ignore already closed error.

Any other error should be surfaced though.

Reply:
So, I made the http server private to prevent it being closed from the outside (independent of the context).
I'm not sure, does this cover your concerns?

Reply:
No, it's an internal problem. It's irrespective from it being closed. :) It internally races and sometimes will return this error.

Reply:
Here is the link: https://pkg.go.dev/net/http#Server.Shutdown

Reply:
I said it incorrectly. It's not a race. It's by design. :)

Reply:
To be honest, I still dont see how that applies to this implementation, since we do not collect the errors from closing listeners during shutdown (which, as I understand, are the things that return the `ErrServerClosed` error).

Reply:
I've already used this within tests  that also check that `Start(...)` does not throw an error. If the `ErrServerClosed` of the listeners would be forwarded, I'd expect these tests to fail every time since, as you pointed out, the `ErrServerClosed` is by design.
I also debugged the code, while `serverErr` contains `ErrServerClosed`, the `err` returned by `Shutdown(...)` is `nil`.